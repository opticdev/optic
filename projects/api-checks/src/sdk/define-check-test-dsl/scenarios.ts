import { OpenAPIV3 } from 'openapi-types';
const HttpMethods = OpenAPIV3.HttpMethods;
type ScenarioName = string;
export type BeforeAndAfter = [
  OpenAPIV3.Document,
  OpenAPIV3.Document,
  ScenarioName
];
export const copy = (obj: any) => JSON.parse(JSON.stringify(obj));
export type Editable<T> = (item: T) => T;

const defaultEmpty = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {},
});

const defaultWithoutSchema = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithResponseBodySchema = (
  schema: OpenAPIV3.SchemaObject | undefined
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
            content: {
              'application/json': {
                schema: schema,
              },
            },
          },
        },
      },
    },
  },
});

const defaultWithRequestParameters = (
  parameters: OpenAPIV3.ParameterObject[]
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        parameters,
        responses: {
          '200': {
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithResponseHeaders = (
  key: string,
  header: OpenAPIV3.HeaderObject,
  include: boolean
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            headers: include
              ? {
                  [key]: header,
                }
              : {},
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithRequestBodySchema = (
  schema: OpenAPIV3.SchemaObject | undefined
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      post: {
        requestBody: {
          content: {
            'application/json': {
              schema,
            },
          },
        },
        responses: {
          '201': {
            description: '',
          },
        },
      },
    },
  },
});

const baseParam = { in: 'query', name: 'exampleParam' };

export function scenario(name: string) {
  return {
    // responseSchema: {
    //   added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
    //     return [defaultWithoutSchema(), defaultWithResponseBodySchema(schema)];
    //   },
    //   changed: (
    //     beforeSchema: OpenAPIV3.SchemaObject,
    //     afterSchema: OpenAPIV3.SchemaObject
    //   ): BeforeAndAfter => {
    //     return [
    //       defaultWithResponseBodySchema(beforeSchema),
    //       defaultWithResponseBodySchema(afterSchema),
    //     ];
    //   },
    // },
    paths: {
      changed: (
        pathsBefore: OpenAPIV3.PathsObject,
        editPaths: Editable<OpenAPIV3.PathsObject>
      ): BeforeAndAfter => {
        return [
          { ...defaultEmpty(), paths: pathsBefore },
          { ...defaultEmpty(), paths: editPaths(copy(pathsBefore)) },
          name,
        ];
      },
    },
    operation: {
      requirement: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const spec: OpenAPIV3.Document = defaultEmpty();
        spec.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [spec, spec, name];
      },
      added: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        copied.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [defaultEmpty(), copied, name];
      },
      removed: (
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        delete copied.paths[pathPattern]?.[method];
        return [copied, defaultEmpty(), name];
      },
      changed: (
        operationBefore: OpenAPIV3.OperationObject,
        operationAfter: (
          before: OpenAPIV3.OperationObject
        ) => OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const before: OpenAPIV3.Document = defaultEmpty();
        before.paths = {
          [pathPattern]: {
            [method]: operationBefore,
          },
        };
        const after: OpenAPIV3.Document = defaultEmpty();
        after.paths = {
          [pathPattern]: {
            [method]: operationAfter(copy(operationBefore)),
          },
        };
        return [before, after, name];
      },
    },
    requestParameter: {
      added: (parameterToAdded: OpenAPIV3.ParameterObject): BeforeAndAfter => {
        return [
          defaultWithRequestParameters([]),
          defaultWithRequestParameters([parameterToAdded]),
          name,
        ];
      },
      changed: (
        parameterBefore: OpenAPIV3.ParameterObject,
        editParameter: Editable<OpenAPIV3.ParameterObject>
      ): BeforeAndAfter => {
        return [
          defaultWithRequestParameters([parameterBefore]),
          defaultWithRequestParameters([editParameter(copy(parameterBefore))]),
          name,
        ];
      },
    },
    responseHeader: {
      added: (key: string, header: OpenAPIV3.HeaderObject): BeforeAndAfter => {
        return [
          defaultWithResponseHeaders(key, header, false),
          defaultWithResponseHeaders(key, header, true),
          name,
        ];
      },
      changed: (
        key: string,
        headerBefore: OpenAPIV3.HeaderObject,
        editHeader: Editable<OpenAPIV3.HeaderObject>
      ): BeforeAndAfter => {
        return [
          defaultWithResponseHeaders(key, headerBefore, true),
          defaultWithResponseHeaders(key, editHeader(copy(headerBefore)), true),
          name,
        ];
      },
    },
    requestBodySchema: {
      changed: (
        schemaBefore: OpenAPIV3.SchemaObject,
        editSchema: Editable<OpenAPIV3.SchemaObject>
      ): BeforeAndAfter => {
        return [
          defaultWithRequestBodySchema(schemaBefore),
          defaultWithRequestBodySchema(editSchema(copy(schemaBefore))),
          name,
        ];
      },
      added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
        return [
          defaultWithRequestBodySchema(undefined),
          defaultWithRequestBodySchema(schema),
          name,
        ];
      },
    },
    responseBodySchema: {
      changed: (
        schemaBefore: OpenAPIV3.SchemaObject,
        editSchema: Editable<OpenAPIV3.SchemaObject>
      ): BeforeAndAfter => {
        return [
          defaultWithResponseBodySchema(schemaBefore),
          defaultWithResponseBodySchema(editSchema(copy(schemaBefore))),
          name,
        ];
      },
      added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
        return [
          defaultWithResponseBodySchema(undefined),
          defaultWithResponseBodySchema(schema),
          name,
        ];
      },
    },
  };
}
