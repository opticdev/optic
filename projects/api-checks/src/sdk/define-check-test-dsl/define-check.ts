import { BeforeAndAfter } from './scenarios';
import { ApiCheckService } from '../api-check-service';
import { ApiChangeDsl } from '../api-change-dsl';
import { factsToChangelog } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export interface IApiCheckDefinition<CheckConfig> {
  name: string;
  implementation?: (apiChangeDsl: ApiChangeDsl, config: CheckConfig) => void;
  description?: string;
  validExamples: BeforeAndAfter[];
  invalidExamples: BeforeAndAfter[];
}

export class ApiCheckImpl<CheckConfig> {
  public check: IApiCheckDefinition<CheckConfig>;

  constructor(name: string) {
    this.check = {
      name,
      validExamples: [],
      invalidExamples: [],
    };
  }

  description(description: string) {
    this.check.description = description;
    return this;
  }

  implementation(
    implementation: (
      apiChangeDsl: ApiChangeDsl,
      config?: CheckConfig
    ) => void | Promise<void>
  ) {
    this.check.implementation = implementation;
    return this;
  }

  runner() {
    return (dsl: ApiChangeDsl) =>
      this.check.implementation!(dsl, undefined as any);
  }

  runnerWithConfig(config: CheckConfig) {
    return (dsl: ApiChangeDsl) => this.check.implementation!(dsl, config);
  }

  passingExample(beforeAndAfter: BeforeAndAfter, config?: CheckConfig) {
    this.check.validExamples = [...this.check.validExamples, beforeAndAfter];

    if (!this.check.implementation) test.skip('', () => {});

    if (
      typeof jest !== 'undefined' &&
      this.check.implementation &&
      process.env['RUN_CHECK_JEST_TESTS']
    ) {
      // @ts-ignore
      const { test, expect } = global;

      const testName = `passing case ${this.check.validExamples.length}: ${beforeAndAfter[2]}`;

      test(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter, config);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeTruthy();
      });
    }

    return this;
  }

  failingExample(beforeAndAfter: BeforeAndAfter, config?: CheckConfig) {
    this.check.invalidExamples = [
      ...this.check.invalidExamples,
      beforeAndAfter,
    ];

    if (!this.check.implementation) test.skip('', () => {});

    if (
      typeof jest !== 'undefined' &&
      this.check.implementation &&
      process.env['RUN_CHECK_JEST_TESTS']
    ) {
      // @ts-ignore
      const { it, expect } = global;

      const testName = `failing case ${this.check.invalidExamples.length}: ${beforeAndAfter[2]}`;
      it(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter, config);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeFalsy();
      });
    }

    return this;
  }

  private async testExample(
    beforeAndAfter: BeforeAndAfter,
    config?: CheckConfig
  ) {
    invariant(this.check.implementation);
    const service = new ApiCheckService().useRulesFrom((dsl) =>
      this.check.implementation!(dsl, config as any)
    );

    const [before, after] = beforeAndAfter;
    const { currentFacts, nextFacts } = service.generateFacts(before, after);
    const runResults = await service.runRulesWithFacts({
      context: {},
      nextFacts,
      currentFacts,
      changelog: factsToChangelog(currentFacts, nextFacts),
      nextJsonLike: after,
      currentJsonLike: before,
    });

    return {
      passed: !runResults.some((i) => !i.passed),
      results: runResults,
    };
  }
}

export function check<CheckConfig = undefined>(name: string) {
  return new ApiCheckImpl<CheckConfig>(name);
}
