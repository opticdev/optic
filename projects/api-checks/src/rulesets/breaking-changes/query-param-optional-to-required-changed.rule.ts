import { check } from '../../sdk/define-check-test-dsl/define-check';
import { scenario } from '../../sdk/define-check-test-dsl/scenarios';

export default check('prevent changing from optional to required')
  .implementation(({ request }) => {
    const { expect } = require('chai');
    request.queryParameter.changed.must(
      'not allow changing from optional to required',
      (beforeParam, afterParam) => {
        if (!beforeParam.required && afterParam.required) {
          expect.fail('expected query parameter to be not be required');
        }
      }
    );
  })
  .failingExample(
    scenario('changing from optional to required').requestParameter.changed(
      {
        in: 'query',
        name: 'exampleParam',
        required: false,
      },
      (param) => {
        param.required = true;
        return param;
      }
    )
  );
