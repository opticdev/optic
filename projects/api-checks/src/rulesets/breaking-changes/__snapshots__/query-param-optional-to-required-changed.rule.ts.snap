// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failing case 1: changing from optional to required 1`] = `
Object {
  "passed": false,
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "in": "query",
            "name": "exampleParam",
            "required": true,
          },
          "before": Object {
            "in": "query",
            "name": "exampleParam",
            "required": false,
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "inRequest": Object {
              "query": "exampleParam",
            },
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "parameters",
            "query",
            "exampleParam",
          ],
          "jsonPath": "/paths/~1example/get/parameters/0",
          "kind": "query-parameter",
        },
      },
      "condition": "not allow changing from optional to required",
      "docsLink": undefined,
      "effectiveOnDate": undefined,
      "error": "expected query parameter to be not be required",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated query-parameter: exampleParam",
    },
  ],
}
`;
